WooCommerce/Classes/ViewRelated/Products/Edit Product/Product Settings/Visibility/ProductVisibilityViewController.swift

```diff
@@ -0,0 +1,306 @@
+import UIKit
+import Yosemite
+
+final class ProductVisibilityViewController: UIViewController {
+
+    @IBOutlet weak private var tableView: UITableView!
+
+    private var sections: [Section] = []
+
+    // Completion callback
+    //
+    typealias Completion = (_ productSettings: ProductSettings) -> Void
+    private let onCompletion: Completion
+
+    private let productSettings: ProductSettings
+
+    private var visibility: ProductVisibility = .publicVisibility
+
+    private lazy var keyboardFrameObserver: KeyboardFrameObserver = {
+        let keyboardFrameObserver = KeyboardFrameObserver { [weak self] keyboardFrame in
+            self?.handleKeyboardFrameUpdate(keyboardFrame: keyboardFrame)
+        }
+        return keyboardFrameObserver
+    }()
+
+    /// Init
+    ///
+    init(settings: ProductSettings, completion: @escaping Completion) {
+        productSettings = settings
+        onCompletion = completion
+        super.init(nibName: nil, bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureNavigationBar()
+        configureMainView()
+        configureTableView()
+        visibility = getProductVisibility(productSettings)
+        reloadSections()
+        handleSwipeBackGesture()
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+
+        /// if the visibility selected is different from protected by password, the password becomes always an empty string (no password)
+        if visibility != .passwordProtected {
+            productSettings.password = ""
+        }
+
+        onCompletion(productSettings)
```

Comment:
I think Jayson mentioned in a PR before, maybe the completion logic can be only under `isMovingFromParent` in `viewWillDisappear` in case it's called from presenting some modal or other unexpected scenarios?

```
if isMovingFromParent {
  // completion
}
```